{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Transcription Factor DNA binding assay data analysis pipeline \n",
    "\n",
    "### Version: SGE version (2/28/2018)\n",
    "\n",
    "##### Niu Du (dniu [at] jcvi.org)\n",
    "##### J. Craig Venter Institute (JCVI)\n",
    "##### La Jolla, CA USA"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### require installation of following tools\n",
    "* fastqc\n",
    "* multiqc\n",
    "* GEM - http://groups.csail.mit.edu/cgs/gem/\n",
    "* MACS2 - https://github.com/taoliu/MACS\n",
    "* bwa\n",
    "* Samtools\n",
    "* Deeptools - https://deeptools.readthedocs.io/en/develop/\n",
    "* meme suite - http://meme-suite.org/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from Bio import SeqIO\n",
    "import re, os\n",
    "from tqdm import tqdm\n",
    "import itertools\n",
    "from matplotlib_venn import venn3\n",
    "from matplotlib.offsetbox import (TextArea, DrawingArea, OffsetImage,AnnotationBbox)\n",
    "from matplotlib.cbook import get_sample_data\n",
    "import matplotlib.image as mpimg\n",
    "\n",
    "#import customed function\n",
    "from Dapseq_functions import Tandem_filter\n",
    "\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#0 Init varibles - edit here\n",
    "os.environ['file_path'] = './Data/DAP_sequence_2016+2018'\n",
    "os.environ['control'] = 'Halovector_S35_L002_R1_001'\n",
    "os.environ['project_code'] = '0746'\n",
    "\n",
    "# Number of threads to use if possible\n",
    "THREADS = 8 \n",
    "os.environ['THREADS'] = str(THREADS)\n",
    "\n",
    "os.environ['gempath'] = '/home/dniu/tools/gem' #java support required \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#0.1 Create index for reference genome - Only need to do this once if not done yet\n",
    "!qsub -P $project_code -N index_genome -cwd bwa index -a bwtsw Genome_REF/PT_Chr_only.fa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#0.2 Create index for reference genome and mask transposon regions\n",
    "'''This step may not be needed'''\n",
    "!/home/dniu/tools/convert2bed/convert2bed -i gff -o bed <MASK-REFERENCE.gff> MASK-REFERENCE.bed\n",
    "!bedtools maskfasta -fi YOUR-REFERENCE-GENOME.fa -bed MASK-REFERENCE.bed -fo MASK-REFERENCE.fa\n",
    "## Also need to modify genome reference path in step 7.2 and 9 depending on the purpose of use"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#1 Trim raw fastq files \n",
    "File_list = os.listdir(os.environ['file_path']+'/')\n",
    "!mkdir -p $file_path/Trimmed_output\n",
    "for file in tqdm(File_list):\n",
    "    if file.endswith(\".gz\"):\n",
    "        os.environ['input_name'] = os.path.join(os.environ['file_path'], file)\n",
    "        os.environ['output_name'] = os.path.join(os.environ['file_path']+\"/Trimmed_output\", file)\n",
    "        !qsub -P $project_code -N trim - cwd -e gridout/err -o gridout/out java -jar /home/jiyindna/Desktop/Dubiosys/Genetools/Trimmomatic-0.36/trimmomatic-0.36.jar  \\\n",
    "        SE -phred33 $input_name $output_name \\\n",
    "        ILLUMINACLIP:TruSeq3-SE:2:30:10 LEADING:3 TRAILING:3 SLIDINGWINDOW:4:15 MINLEN:36 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#1.1 Getting sample names for downstream processes\n",
    "File_list = os.listdir(os.environ['file_path']+\"/Trimmed_output/\")\n",
    "Sample_list = [item.split('.')[0] for item in File_list if item.endswith(\".gz\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#1.2 In this step (1.2) and the following step (1.3), you can manualy delete some of the samples that you do not wish to analyse\n",
    "with open(os.environ['file_path']+\"/Trimmed_output/filelist.txt\",\"w\") as f:\n",
    "    for item in Sample_list:\n",
    "        #if item.endswith(\".gz\"):\n",
    "        f.write(\"%s\\n\" % item)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#1.3 Edit your file then run this step \n",
    "with open(os.environ['file_path']+\"/Trimmed_output/filelist.txt\",\"r\") as f:\n",
    "        Sub_sample_list = [item.split('\\n')[0] for item in tqdm(f)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#2 QC the trimmmed files\n",
    "!mkdir -p ./$project_ID/Trimmed_output/Fastqc\n",
    "for item in File_list:\n",
    "    os.environ['item'] = item\n",
    "    if item.endswith(\".gz\"):\n",
    "        !qsub -P $project_code -N trim -cwd -e gridout/err -o gridout/out /home/dniu/tools/FastQC/fastqc ./$project_ID/Trimmed_output/$item \\\n",
    "        --outdir=./$project_ID/Trimmed_output/Fastqc/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 2.1 combine all the fastqc report and merge into one multiqc report\n",
    "!multiqc ./$file_path/Trimmed_output/Fastqc \n",
    "!zip -r $file_path.zip multiqc_report.html ./multiqc_data\n",
    "!rm -r ./multiqc_data\n",
    "!rm multiqc_report.html"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#3 Seq alignment and write sam files\n",
    "# https://arxiv.org/abs/1303.3997\n",
    "\n",
    "#Important parameters for tuning\n",
    "# -k Minimum seed length. Matches shorter than INT will be missed.\n",
    "# -B Mismatch penalty. The sequence error rate is approximately: {.75 * exp[-log(4) * B/A]}\n",
    "# -O Gap open penalty.[6]\n",
    "\n",
    "!mkdir -p $file_path/Sam_File\n",
    "for item in Sub_sample_list:\n",
    "    os.environ['sample_ID'] = item   \n",
    "    # Note bwa mem does not take masked genome\n",
    "    !qsub -P $project_code -N BWA.$sample_ID -cwd -pe threaded $THREADS -e gridout/err -o gridout/out \\\n",
    "    ./run_bwa_mem.py -t $THREADS -k 60 -B 7 -O 6 -v 0 -r $path/Genome_REF/PT_Chr_only.fa \\\n",
    "    -i $file_path/Trimmed_output/$sample_ID.fastq.gz -o $file_path/Sam_File/$sample_ID.sam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#4 filter out unmapped reads and sort mapped reads\n",
    "!mkdir -p $file_path/Sam_sorted\n",
    "for item in Sub_sample_list:\n",
    "    os.environ['sample_ID'] = item\n",
    "    !qsub -P $project_code -N SORT.$sample_ID -hold_jid BWA.$sample_ID -cwd -pe threaded $THREADS -e gridout/err -o gridout/out \\\n",
    "    ./run_Sam_view_sort.py -t $THREADS -i $file_path/Sam_File/$sample_ID.sam -o $file_path/Sam_sorted/$sample_ID._temp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Continue 4 - run after the previous step is finished\n",
    "for item in tqdm(Sub_sample_list):\n",
    "    os.environ['sample_ID'] = item\n",
    "    !samtools rmdup -s $file_path/Sam_sorted/$sample_ID._temp.bam $file_path/Sam_sorted/$sample_ID._mapped_sorted.bam \n",
    "    !samtools index $file_path/Sam_sorted/$sample_ID._mapped_sorted.bam "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#4.1 Vis singals using all mapped reads - Visulization step\n",
    "!mkdir -p $file_path/bigWig\n",
    "for item in Sub_sample_list:\n",
    "    os.environ['sample_ID'] = item\n",
    "    !qsub -P $project_code -N bamCoverage -cwd -e gridout/err -o gridout/out \\\n",
    "    bamCoverage -b $file_path/Sam_sorted/$sample_ID._mapped_sorted.bam \\\n",
    "    --normalizeUsing BPM \\\n",
    "    --extendReads 300 \\\n",
    "    -o $file_path/bigWig/$sample_ID.bw \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#4.2 Compare bam files and target peak regions\n",
    "for item in Sub_sample_list:\n",
    "    if os.environ['control'] not in item:\n",
    "        os.environ['sample_ID'] = item\n",
    "        !qsub -P $project_code -N bamCompare -cwd -e gridout/err -o gridout/out \\\n",
    "        bamCompare  -b1 $file_path/Sam_sorted/$sample_ID._mapped_sorted.bam \\\n",
    "        -b2 $file_path/Sam_sorted/$control._mapped_sorted.bam \\\n",
    "        -o $file_path/bigWig/$sample_ID.peaks.bw \\\n",
    "        --binSize 80 --operation ratio --scaleFactorsMethod SES -n 1000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#5 MACS2 for peak calling\n",
    "#-B store all details\n",
    "# -q/--qvalue The qvalue (minimum FDR) cutoff to call significant regions.\n",
    "'''Default is 0.05. For broad marks, you can try 0.01 as cutoff. \n",
    "Q-values are calculated from p-values using Benjamini-Hochberg procedure.'''\n",
    "\n",
    "for item in tqdm(Sub_sample_list):\n",
    "    if os.environ['control'] not in item:\n",
    "        os.environ['sample_ID'] = item\n",
    "        !/u01/dniu/anaconda3/bin/macs2 callpeak -t $file_path/Sam_sorted/$sample_ID._mapped_sorted.bam \\\n",
    "        -c $file_path/Sam_sorted/$control._mapped_sorted.bam \\\n",
    "        -f BAM --outdir $file_path/MACS -g 1.6e7 -n $sample_ID -B -q 0.01 -m 2 50 --verbose=0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#6 GEM tool for peak calling and motif model creation - use BAM file\n",
    "# force q value to be 0.001, q = 3, 0.01, q =2\n",
    "# Only to use GPS data here\n",
    "\n",
    "# The p-value of the motif occurence. \n",
    "'''The p-value is the probability of a random sequence of the same \n",
    "length as the motif matching that position of the sequence with a \n",
    "score at least as good.'''\n",
    "\n",
    "# The q-vlavlue of the motif occurence. \n",
    "'''The q-value is the estimated false discovery rate if the occurrence \n",
    "is accepted as significant.''' \n",
    "\n",
    "#The width (bp) to smooth the read distribution. \n",
    "'''If it is set to -1, there will be no smoothing (default=30). \n",
    "For ChIP-exo, use a smaller number (say 3 or 5) to achieve higher spatial accuracy.'''\n",
    "\n",
    "# Fold (IP/Control) cutoff to filter predicted events (default=3)\n",
    "# Depending on the sequencing depth, the sample read\n",
    "\n",
    "# Proc. Natl Acad. Sci. USA (2003) 100:9440â€“9445\n",
    "!rm -r $file_path/GEM_BED\n",
    "!mkdir $file_path/GEM_BED\n",
    "for item in Sub_sample_list:\n",
    "    if os.environ['control'] not in item:\n",
    "        os.environ['sample_ID'] = item\n",
    "        !qsub -P $project_code -N GEM.$sample_ID -cwd -pe threaded $THREADS -e gridout/err -o gridout/out \\\n",
    "        java -jar -Xmx16G $gempath/gem.jar --t $THREADS \\\n",
    "        --f BAM \\\n",
    "        --d $gempath/Read_Distribution_default.txt \\\n",
    "        --g ./Genome_REF/PT_genome_size.txt \\\n",
    "        --genome ./Genome_REF/PT_Chr \\\n",
    "        --expt $file_path/Sam_sorted/$sample_ID._mapped_sorted.bam \\\n",
    "        --ctrl $file_path/Sam_sorted/$control._mapped_sorted.bam \\\n",
    "        --outBED \\\n",
    "        --out $file_path/GEM_BED/$sample_ID \\\n",
    "        --k_min 6 --kmax 20 --k_seqs 600 --k_neg_dinu_shuffle\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "########## Choose one of the selected option"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#7 Combining peak regions from MACS2 and GEM\n",
    "\n",
    "# Set lower limit for MACS2 output\n",
    "min_score = 1\n",
    "window_size = 80\n",
    "!mkdir -p $file_path/Combined_BED\n",
    "for item in tqdm(Sub_sample_list):\n",
    "    if os.environ['control'] not in item:\n",
    "        os.environ['sample_ID'] = item\n",
    "        # Read in bed files from MACS2 output\n",
    "        try:\n",
    "            df_bed = pd.read_table(os.environ['file_path']+'/MACS/'+item+'_summits.bed',header = None)\n",
    "            chrs = list(df_bed[df_bed[4]>min_score][0])\n",
    "            peaks = list(df_bed[df_bed[4]>min_score][1])\n",
    "            names = ['MACS_'+str(ix) for ix in range(len(df_bed[df_bed[4]>min_score]))]\n",
    "            scores = list(df_bed[df_bed[4]>min_score][4])\n",
    "\n",
    "        # Read in location files from GEM output and add to the list\n",
    "\n",
    "            df_GEM = pd.read_table(os.environ['file_path']+'/GEM_BED/'+item+'/'+item+'.GEM_events.txt')\n",
    "            chrs = chrs + ['chr'+pos.split(':')[0] for pos in df_GEM[df_GEM['Fold']>min_score].Position]\n",
    "            peaks = peaks + [int(pos.split(':')[1]) for pos in df_GEM[df_GEM['Fold']>min_score].Position]\n",
    "            names = names + ['GEM_'+str(ix) for ix in range(len(df_GEM[df_GEM['Fold']>min_score]))]\n",
    "            scores = scores + list(df_GEM[df_GEM['Fold']>min_score].Fold)   \n",
    "            \n",
    "            with open(os.environ['file_path']+'/Combined_BED/'+item+'.combined.bed','w') as handle:\n",
    "                for c,p,n,s in zip(chrs,peaks,names,scores):\n",
    "                    handle.write(c+'\\t'+str(p-int(window_size/2))+'\\t'+str(p+int(window_size/2))+'\\t'+n+'\\t'+str(s)+'\\n')\n",
    "\n",
    "        except:\n",
    "            continue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#7.1 Generate bed files based on the output from GEM and MACS2, and compare the intercets \n",
    "min_score = 1\n",
    "window_size = 80\n",
    "\n",
    "# Get all valid files from previous step\n",
    "Sub_sample_list1 = [name.split('.')[0] for name in os.listdir(os.environ['file_path']+\"/Combined_BED/\") if name.endswith('bed')]\n",
    "\n",
    "!mkdir -p $file_path/compare_bed\n",
    "for item in tqdm(Sub_sample_list1):\n",
    "    if os.environ['control'] not in item:\n",
    "        os.environ['sample_ID'] = item\n",
    "        \n",
    "        df_bed = pd.read_table(os.environ['file_path']+'/MACS/'+item+'_summits.bed',header = None)\n",
    "        chrs = list(df_bed[df_bed[4]>min_score][0])\n",
    "        peaks = list(df_bed[df_bed[4]>min_score][1])\n",
    "        names = ['MACS_'+str(ix) for ix in range(len(df_bed[df_bed[4]>min_score]))]\n",
    "        scores = list(df_bed[df_bed[4]>min_score][4])\n",
    "        \n",
    "        with open(os.environ['file_path']+'/compare_bed/'+item+'.MACS.bed','w') as handle:\n",
    "            for c,p,n,s in zip(chrs,peaks,names,scores):\n",
    "                handle.write(c+'\\t'+str(p-int(window_size/2))+'\\t'+str(p+int(window_size/2))+'\\t'+n+'\\t'+str(s)+'\\n')\n",
    "        \n",
    "        df_GEM = pd.read_table(os.environ['file_path']+'/GEM_BED/'+item+'/'+item+'.GEM_events.txt')\n",
    "        chrs = ['chr'+pos.split(':')[0] for pos in df_GEM[df_GEM['Fold']>min_score].Position]\n",
    "        peaks = [int(pos.split(':')[1]) for pos in df_GEM[df_GEM['Fold']>min_score].Position]\n",
    "        names = ['GEM_'+str(ix) for ix in range(len(df_GEM[df_GEM['Fold']>min_score]))]\n",
    "        scores = list(df_GEM[df_GEM['Fold']>min_score].Fold)\n",
    "        \n",
    "        with open(os.environ['file_path']+'/compare_bed/'+item+'.GEM.bed','w') as handle:\n",
    "            for c,p,n,s in zip(chrs,peaks,names,scores):\n",
    "                handle.write(c+'\\t'+str(p-int(window_size/2))+'\\t'+str(p+int(window_size/2))+'\\t'+n+'\\t'+str(s)+'\\n')\n",
    "    \n",
    "        \n",
    "        !bedtools intersect -wo -a $file_path/compare_bed/$sample_ID.MACS.bed -b $file_path/compare_bed/$sample_ID.GEM.bed > $file_path/compare_bed/$sample_ID.compare.bed "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#7.2 Generate fasta files from bed files and remove duplicated sequences\n",
    "!mkdir -p $file_path/fasta\n",
    "for item in tqdm(Sub_sample_list1):\n",
    "    if os.environ['control'] not in item:\n",
    "        os.environ['sample_ID'] = item\n",
    "        os.environ['fasta_file'] = os.environ['file_path']+'/fasta/'+os.environ['sample_ID']+'.fasta'\n",
    "        !bedtools getfasta -fo $fasta_file -fi ./Genome_REF/PT_Chr_only.fa \\\n",
    "        -bed $file_path/Combined_BED/$sample_ID.combined.bed\n",
    "        \n",
    "        key_list = []\n",
    "        dict_fasta = {}\n",
    "        with open(os.environ['fasta_file'],'r') as handle:\n",
    "            with open(os.environ['fasta_file']+'.nodup','w') as output_handle:\n",
    "                for rec in SeqIO.parse(handle,'fasta'):\n",
    "                    if rec.name not in key_list:\n",
    "                        key_list.append(rec.name)\n",
    "                        output_handle.write('>'+rec.name+'\\n'+str(rec.seq)+'\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#7.3 Filter out tandem sequences \n",
    "\n",
    "# P of 6-mer show up in n base pair\n",
    "# m time n/(4^6)^m - 1 time 0.12 -2 times 0.3e-3 -3 times 7.3e-9\n",
    "for count,item in enumerate(Sub_sample_list1):\n",
    "    if os.environ['control'] not in item:\n",
    "        os.environ['sample_ID'] = item\n",
    "        print(str(count+1)+'/'+str(len(Sub_sample_list1)),item)\n",
    "        fasta_file = os.environ['file_path']+'/fasta/'+os.environ['sample_ID']+'.fasta.nodup'\n",
    "        Tandem_filter(fasta_file,k = 6,k_max = 4)    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#8 Use meme suite to build motif model\n",
    "\n",
    "'''-objfun ce - This objective function scores motifs based on their enrichment in the central regions of the primary sequences, which must all be of same the length. (See option -cefrac, below.) \n",
    "The width and number of sites of each motif are determined using the binomial test. '''\n",
    "# -mod anr/oops/zoops\n",
    "# -ws The gap extension cost for creating the alignments used to trim the motif.\n",
    "\n",
    "for item in Sub_sample_list1:\n",
    "    if os.environ['control'] not in item:\n",
    "        os.environ['sample_ID'] = item\n",
    "        !qsub -P $project_code -N MEME.$sample_ID -cwd -e gridout/err -o gridout/out \\\n",
    "        meme $file_path/fasta/$sample_ID.fasta.nodup \\\n",
    "        -nmotifs 1 -minw 4 -maxw 12 -dna -mod oops -nostatus \\\n",
    "        -oc $file_path/GEM_BED/$sample_ID-meme"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#9 Locate motif loci in the genome\n",
    "for item in Sub_sample_list1:\n",
    "    if os.environ['control'] not in item:\n",
    "        os.environ['sample_ID'] = item\n",
    "        !qsub -P $project_code -N FIMO.$sample_ID -hold_jid MEME.$sample_ID -cwd -e gridout/err -o gridout/out \\\n",
    "        fimo -verbosity 1 --thresh 1e-5\\\n",
    "        -oc $file_path/GEM_BED/$sample_ID-fimo/ \\\n",
    "        $file_path/GEM_BED/$sample_ID-meme/meme.txt \\\n",
    "        ./Genome_REF/PT_Chr_only.fa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#9.1 Convert gff to bed\n",
    "Sub_sample_list2 = [name.split('-')[0] for name in os.listdir(os.environ['file_path']+\"/GEM_BED/\") if name.endswith('fimo')]\n",
    "\n",
    "for item in Sub_sample_list2:\n",
    "    if os.environ['control'] not in item:\n",
    "        os.environ['sample_ID'] = item\n",
    "        !qsub -P $project_code -N gff2bed.$sample_ID -hold_jid FIMO.$sample_ID -cwd -e gridout/err \\\n",
    "        -o $file_path/GEM_BED/$sample_ID-fimo/fimo.bed \\\n",
    "        gff2bed -i $file_path/GEM_BED/$sample_ID-fimo/fimo.gff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#9.2 Compare motif to peaks called by GEM and MACS2\n",
    "\n",
    "def get_type(x):\n",
    "    if len(x.split(','))>1:\n",
    "        return 0\n",
    "    elif 'MACS' in x:\n",
    "        return 1\n",
    "    elif 'GEM' in x:\n",
    "        return 2\n",
    "    else:\n",
    "        return -1\n",
    "    \n",
    "venn_set = {}\n",
    "for item in tqdm(Sub_sample_list2):\n",
    "    if os.environ['control'] not in item:\n",
    "        os.environ['sample_ID'] = item\n",
    "        try:\n",
    "            !bedtools intersect -wa -a $file_path/GEM_BED/$sample_ID-fimo/fimo.bed -b $file_path/compare_bed/$sample_ID.MACS.bed \\\n",
    "            > $file_path/compare_bed/$sample_ID.MACS_peak.bed\n",
    "            !bedtools intersect -wa -a $file_path/GEM_BED/$sample_ID-fimo/fimo.bed -b $file_path/compare_bed/$sample_ID.GEM.bed \\\n",
    "            > $file_path/compare_bed/$sample_ID.GEM_peak.bed\n",
    "            !bedtools intersect -wa -a $file_path/compare_bed/$sample_ID.MACS.bed -b $file_path/compare_bed/$sample_ID.GEM.bed \\\n",
    "            > $file_path/compare_bed/$sample_ID.compare_peak.bed\n",
    "            # !bedtools intersect -wa -a $file_path/GEM_BED/$sample_ID-fimo/fimo.bed -b $file_path/compare_bed/$sample_ID.MACS.bed $file_path/compare_bed/$sample_ID.GEM.bed \\\n",
    "            # > $file_path/compare_bed/$sample_ID.all_peak.bed\n",
    "\n",
    "            df_MACS = pd.read_table(os.environ['file_path']+'/compare_bed/'+os.environ['sample_ID']+'.MACS.bed',header = None)\n",
    "            df_GEM = pd.read_table(os.environ['file_path']+'/compare_bed/'+os.environ['sample_ID']+'.GEM.bed',header = None)\n",
    "            df_Motif = pd.read_table(os.environ['file_path']+'/GEM_BED/'+os.environ['sample_ID']+'-fimo/fimo.bed',header = None) \n",
    "\n",
    "            df_compare_MACS_peak = pd.read_table(os.environ['file_path']+'/compare_bed/'+os.environ['sample_ID']+'.MACS_peak.bed',header = None)\n",
    "            #df_compare_MACS_peak = df_compare_MACS_peak.drop_duplicates(7)\n",
    "            df_compare_GEM_peak = pd.read_table(os.environ['file_path']+'/compare_bed/'+os.environ['sample_ID']+'.GEM_peak.bed',header = None)\n",
    "            #df_compare_GEM_peak = df_compare_GEM_peak.drop_duplicates(7)\n",
    "            df_compare = pd.read_table(os.environ['file_path']+'/compare_bed/'+os.environ['sample_ID']+'.compare_peak.bed',header = None)\n",
    "\n",
    "            # compare all three \n",
    "            df_compare_motif = pd.read_table(os.environ['file_path']+'/compare_bed/'+os.environ['sample_ID']+'.all_peak.bed',header = None)\n",
    "            df_compare_motif = df_compare_motif.drop_duplicates()\n",
    "            \n",
    "            All_overlay = len(set(df_compare_MACS_peak[3]).intersection(df_compare_GEM_peak[3]))\n",
    "            \n",
    "            # Unique to MACS\n",
    "            MACS_only = len(df_MACS)-len(df_compare)-len(df_compare_MACS_peak)+All_overlay\n",
    "            # Unique to GEM\n",
    "            GEM_only = len(df_GEM)-len(df_compare)-len(df_compare_GEM_peak)+All_overlay\n",
    "            # Unique to MOTIF\n",
    "            MOTIF_only = len(df_Motif) - len(df_compare_MACS_peak) - len(df_compare_GEM_peak) + All_overlay\n",
    "\n",
    "            # GEM overlap MACS only\n",
    "            GEM_MACS_only = len(df_compare) - All_overlay\n",
    "            # MACS overlap MOTIF only\n",
    "            MACS_MOTIF_only = len(df_compare_MACS_peak) - All_overlay\n",
    "            # GEM overlap MOTIF only\n",
    "            GEM_MOTIF_only = len(df_compare_GEM_peak) - All_overlay\n",
    "            \n",
    "            # Get the overaly motif and write into bed format\n",
    "            df_Motif.set_index(3).loc[set(df_compare_MACS_peak[3]).union(df_compare_GEM_peak[3])].reset_index()[[0,1,2,3]].to_csv(os.environ['file_path']+'/Combined_BED/'+os.environ['sample_ID']+'.intersection.bed',header = None, index = False, sep = '\\t')\n",
    "            venn_set[item] = [MACS_only, GEM_only, GEM_MACS_only, MOTIF_only,MACS_MOTIF_only,GEM_MOTIF_only,All_overlay]\n",
    "        except:\n",
    "            continue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#10 Visulize the final results\n",
    "%matplotlib inline\n",
    "f,ax = plt.subplots(2*int(len(venn_set)/4)+2,4,figsize = (24,max(8,len(venn_set)*2)))\n",
    "ax = ax.ravel()\n",
    "[axi.set_axis_off() for axi in ax.ravel()]  \n",
    "for ix,sets in enumerate(venn_set):\n",
    "    ax[2*ix+1].set_title(str(sets))\n",
    "    venn3(subsets = (venn_set[sets]),set_labels = ('MACS', 'GEM', 'Motif'),ax = ax[2*ix+1],)\n",
    "    ax[2*ix].set_title(str(sets))\n",
    "    img = mpimg.imread(os.environ['file_path']+'/GEM_BED/'+sets+'-meme/logo1.png')\n",
    "    ax[2*ix].imshow(img);\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#11 Find motif that its locaiton overlaps with GEM and MACS targeted peaks\n",
    "\n",
    "!mkdir -p $file_path/fasta\n",
    "for item in tqdm(list(venn_set.keys())):\n",
    "    if os.environ['control'] not in item:\n",
    "        os.environ['sample_ID'] = item\n",
    "        os.environ['fasta_file'] = os.environ['file_path']+'/fasta/'+os.environ['sample_ID']+'.motif.fasta'\n",
    "        !bedtools getfasta -fo $fasta_file -fi ./Genome_REF/PT_Chr_only.fa \\\n",
    "        -bed $file_path/Combined_BED/$sample_ID.intersection.bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#12 Do another round of MEME to build better motif model\n",
    "for item in list(venn_set.keys()):\n",
    "    if os.environ['control'] not in item:\n",
    "        os.environ['sample_ID'] = item\n",
    "        !qsub -P $project_code -N MEME2.$sample_ID -cwd -e gridout/err -o gridout/out \\\n",
    "        meme $file_path/fasta/$sample_ID.motif.fasta \\\n",
    "        -nmotifs 1 -minw 4 -maxw 12 -dna -mod oops -nostatus \\\n",
    "        -oc $file_path/GEM_BED/$sample_ID-meme-intersection\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "f,ax = plt.subplots(2*int(len(venn_set)/4)+2,4,figsize = (24,max(8,len(venn_set)*2)))\n",
    "ax = ax.ravel()\n",
    "[axi.set_axis_off() for axi in ax.ravel()]  \n",
    "for ix,sets in enumerate(venn_set):\n",
    "    ax[2*ix+1].set_title(str(sets))\n",
    "    venn3(subsets = (venn_set[sets]),set_labels = ('MACS', 'GEM', 'Motif'),ax = ax[2*ix+1],)\n",
    "    ax[2*ix].set_title(str(sets))\n",
    "    img = mpimg.imread(os.environ['file_path']+'/GEM_BED/'+sets+'-meme-intersection/logo1.png')\n",
    "    ax[2*ix].imshow(img);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#10 Save all useful output files\n",
    "\n",
    "!rm big_Wig.zip\n",
    "!zip -r -q -J -D big_Wig.zip $file_path/GEM_BED/*-fimo/fimo.gff $file_path/bigWig/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!rm motif_locaiton.zip\n",
    "!zip -q -r -D -J motif_locaiton.zip $file_path/Combined_BED/*.intersection.bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "py35",
   "language": "python",
   "name": "py35"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
